FV3/atmos_cubed_sphere/tools/coarse_grained_restart_files.F90

```diff
@@ -0,0 +1,366 @@
+    endif
+  end subroutine coarse_grain_restart_data_on_model_levels
+
+  subroutine coarse_grain_fv_core_restart_data_on_model_levels(Atm, mass)
+    type(fv_atmos_type), intent(inout) :: Atm
+    real, intent(in) :: mass(is:ie,js:je,1:npz)
+
+    call weighted_block_edge_average_x(Atm%gridstruct%dx(is:ie,js:je+1), &
+         Atm%u(is:ie,js:je+1,1:npz), Atm%coarse_graining%restart%u)
+    call weighted_block_edge_average_y(Atm%gridstruct%dy(is:ie+1,js:je), &
+         Atm%v(is:ie+1,js:je,1:npz), Atm%coarse_graining%restart%v)
+
+    if (.not. Atm%flagstruct%hydrostatic) then
+       call weighted_block_average(mass(is:ie,js:je,1:npz), &
+            Atm%w(is:ie,js:je,1:npz), Atm%coarse_graining%restart%w)
+       call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+            Atm%delz(is:ie,js:je,1:npz), Atm%coarse_graining%restart%delz)
+       if (Atm%flagstruct%hybrid_z) then
+          call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+            Atm%ze0(is:ie,js:je,1:npz), Atm%coarse_graining%restart%ze0)
+       endif
+    endif
+
+    call weighted_block_average(mass(is:ie,js:je,1:npz), &
+         Atm%pt(is:ie,js:je,1:npz), Atm%coarse_graining%restart%pt)
+    call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+         Atm%delp(is:ie,js:je,1:npz), Atm%coarse_graining%restart%delp)
+    call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+         Atm%phis(is:ie,js:je), Atm%coarse_graining%restart%phis)
+
+    if (Atm%flagstruct%agrid_vel_rst) then
+       call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+            Atm%ua(is:ie,js:je,1:npz), Atm%coarse_graining%restart%ua)
+       call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+            Atm%va(is:ie,js:je,1:npz), Atm%coarse_graining%restart%va)
+    endif
+  end subroutine coarse_grain_fv_core_restart_data_on_model_levels
+
+  subroutine coarse_grain_fv_tracer_restart_data_on_model_levels(Atm, mass)
+    type(fv_atmos_type), intent(inout) :: Atm
+    real, intent(in) :: mass(is:ie,js:je,1:npz)
+
+    character(len=64) :: tracer_name
+    integer :: n_tracer
+
+    do n_tracer = 1, n_prognostic_tracers
+       call get_tracer_names(MODEL_ATMOS, n_tracer, tracer_name)
+       if (trim(tracer_name) .eq. 'cld_amt') then
+          call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+               Atm%q(is:ie,js:je,1:npz,n_tracer), &
+               Atm%coarse_graining%restart%q(is_coarse:ie_coarse,js_coarse:je_coarse,1:npz,n_tracer))
+       else
+          call weighted_block_average(mass(is:ie,js:je,1:npz), &
+               Atm%q(is:ie,js:je,1:npz,n_tracer), &
+               Atm%coarse_graining%restart%q(is_coarse:ie_coarse,js_coarse:je_coarse,1:npz,n_tracer))
+       endif
+    enddo
+
+    do n_tracer = n_prognostic_tracers + 1, n_tracers
+       call weighted_block_average(mass(is:ie,js:je,1:npz), &
+               Atm%qdiag(is:ie,js:je,1:npz,n_tracer), &
```

Comment:
Does the last index for `Atm%qdiag` really start at `n_prognostic + 1`?

Reply:
Yes, I believe so; again this is modeled after existing code.  See:

https://github.com/VulcanClimateModeling/fv3gfs-fortran/blob/42f2877a4e04882a1046d35662bce61c85385e27/FV3/atmos_cubed_sphere/tools/fv_io.F90#L568-L574